File: elixir/prime-factors/prime_factors.exs
defmodule PrimeFactors do
  @doc """
  Compute the prime factors for 'number'.

  The prime factors are prime numbers that when multiplied give the desired
  number.

  The prime factors of 'number' will be ordered lowest to highest. 
  """
  @spec factors_for(pos_integer) :: [pos_integer]
  def factors_for(number) do
    factors_for(number, 2, [])
  end

  defp factors_for(1, _, prime_factors), do: Enum.reverse(prime_factors)
  defp factors_for(number, prime, prime_factors) do
    if rem(number, prime) == 0 do
      factors_for(div(number, prime), prime, [prime|prime_factors])
    else
      factors_for(number, next_prime(prime), prime_factors)
    end
  end

  defp next_prime(2), do: 3
  defp next_prime(n) do
    Stream.iterate(n + 2, &(&1 + 2))
    |> Stream.drop_while(&(!prime?(&1)))
    |> Enum.take(1)
    |> hd
  end

  defp prime?(n) when is_integer(n) do
    ceiling = round(:math.sqrt(n))
    if n == ceiling * ceiling do
      false
    else
      prime_test(n, 2, ceiling, ceiling)
    end
  end

  defp prime_test(_, i, i, h), do: i == h
  defp prime_test(n, i, j, h) do
    if rem(n, i) == 0 do
      i == h
    else
      prime_test(n, i + 1, j, h)
    end
  end
end

File: go/prime-factors/prime_factors.go
package prime

import "math"

const testVersion = 2

func Factors(n int64) []int64 {
	return factorsFor(n, 2, []int64{})
}

func factorsFor(n, prime int64, primeFactors []int64) []int64 {
	if n == 1 {
		return primeFactors
	}
	if n%prime == 0 {
		return factorsFor(n/prime, prime, append(primeFactors, prime))
	}
	return factorsFor(n, nextPrime(prime), primeFactors)
}

func nextPrime(n int64) int64 {
	if n == 2 {
		return 3
	}
	for n = n + 2; !isPrime(n); n += 2 {
	}
	return n
}

func isPrime(n int64) bool {
	ceiling := int64(math.Sqrt(float64(n)))
	if n == ceiling*ceiling {
		return false
	}
	return primeTest(n, 2, ceiling, ceiling)
}

func primeTest(n, i, j, h int64) bool {
	if i == j {
		return i == h
	}
	if n%i == 0 {
		return i == h
	}
	i++
	return primeTest(n, i, j, h)
}
